File: app/src/main/java/com/perflyst/twire/tasks/GetTwitchEmotesTask.java
package com.perflyst.twire.tasks;

import com.perflyst.twire.TwireApplication;
import com.perflyst.twire.model.Emote;
import com.perflyst.twire.utils.Execute;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import timber.log.Timber;

/**
 * Created by idealMJ on 29/07/16.
 */
public class GetTwitchEmotesTask implements Runnable {
    private final Delegate delegate;
    private final List<String> emoteSets;
    private final List<Emote> twitchEmotes = new ArrayList<>();
    private final List<Emote> subscriberEmotes = new ArrayList<>();

    public GetTwitchEmotesTask(List<String> emoteSets, Delegate delegate) {
        this.emoteSets = emoteSets;
        this.delegate = delegate;
    }

    public void run() {
        var emotes = TwireApplication.helix.getEmoteSets(null, emoteSets).execute().getEmotes();

        for (var emoteData : emotes) {
            Emote emote = Emote.Twitch(emoteData.getName(), emoteData.getId());
            if (Objects.equals(emoteData.getEmoteSetId(), "0")) {
                twitchEmotes.add(emote);
            } else {
                emote.isSubscriberEmote = true;
                subscriberEmotes.add(emote);
            }
        }
        Collections.sort(twitchEmotes);

        Timber.tag("Chat").d("Found twitch emotes: %s", twitchEmotes.size());
        Execute.ui(() -> delegate.onEmotesLoaded(twitchEmotes, subscriberEmotes));
    }

    public interface Delegate {
        void onEmotesLoaded(List<Emote> emotes, List<Emote> subscriberEmotes);
    }
}

File: app/src/main/java/com/perflyst/twire/tasks/GetTwitchEmotesTask.kt
package com.perflyst.twire.tasks

import com.perflyst.twire.TwireApplication.Companion.helix
import com.perflyst.twire.model.Emote
import com.perflyst.twire.model.Emote.Companion.Twitch
import com.perflyst.twire.utils.Execute
import timber.log.Timber

/**
 * Created by idealMJ on 29/07/16.
 */
class GetTwitchEmotesTask(
    private val emoteSets: MutableList<String>,
    private val delegate: Delegate
) : Runnable {
    private val twitchEmotes: MutableList<Emote> = ArrayList()
    private val subscriberEmotes: MutableList<Emote> = ArrayList()

    override fun run() {
        val emotes = emoteSets.chunked(25)
            .flatMap { helix.getEmoteSets(null, it).execute().emotes }

        for (emoteData in emotes) {
            val emote = Twitch(emoteData.name, emoteData.id)
            if (emoteData.emoteSetId == "0") {
                twitchEmotes.add(emote)
            } else {
                emote.isSubscriberEmote = true
                subscriberEmotes.add(emote)
            }
        }
        twitchEmotes.sort()

        Timber.tag("Chat").d("Found twitch emotes: %s", twitchEmotes.size)
        Execute.ui { delegate.onEmotesLoaded(twitchEmotes, subscriberEmotes) }
    }

    fun interface Delegate {
        fun onEmotesLoaded(emotes: MutableList<Emote>?, subscriberEmotes: MutableList<Emote>?)
    }
}
